src/middlewared/middlewared/plugins/disk_/configure_swap.py

```diff
@@ -0,0 +1,223 @@
+import os
+import platform
+import subprocess
+
+from collections import defaultdict
+
+from middlewared.service import CallError, private, Service
+from middlewared.utils import Popen, run
+
+IS_LINUX = platform.system().lower() == 'linux'
+MIRROR_MAX = 5
+
+
+class DiskService(Service):
+
+    @private
+    async def swaps_configure(self):
+        """
+        Configures swap partitions in the system.
+        We try to mirror all available swap partitions to avoid a system
+        crash in case one of them dies.
+        """
+        used_partitions_in_mirror = set()
+        create_swap_devices = {}
+        disks = [i async for i in await self.middleware.call('pool.get_disks')]
+        disks.extend(await self.middleware.call('boot.get_disks'))
+        existing_swap_devices = {'mirrors': [], 'partitions': []}
+        mirrors = await self.middleware.call('disk.get_mirrors')
+        encrypted_mirrors = {m['encrypted_provider']: m for m in mirrors if m['encrypted_provider']}
+        all_partitions = {p['name']: p for p in await self.middleware.call('disk.list_all_partitions')}
+
+        for device in await self.middleware.call('disk.get_swap_devices'):
+            if device in encrypted_mirrors or device.startswith(('/dev/md', 'mirror/')):
+                # This is going to be complete path for linux and mirror/swapname.eli for freebsd
+                existing_swap_devices['mirrors'].append(device)
+            else:
+                existing_swap_devices['partitions'].append(device)
+
+        # disk.get_mirrors is going to get us complete path for linux and mirror/swapname for freebsd
+        # point to note is that we don't get .eli suffix from above - so we have to be careful here and
+        # use encrypted_provider instead
+        for mirror in mirrors:
+            # If the mirror is degraded or disk is not in a pool lets remove it
+            mirror_name = (mirror['encrypted_provider'] or mirror['real_path']) if IS_LINUX else (
+                mirror['encrypted_provider'] or mirror['path']
+            ).strip('/dev/')
+            swap_mirror = mirror['name'].split(':')[-1].startswith('swap') if IS_LINUX else (
+                mirror['name'].startswith('swap') or mirror['name'].endswith('.sync')
+            )
+            if swap_mirror and (len(mirror['providers']) == 1 or any(
+                p['disk'] not in disks for p in mirror['providers']
+            )):
+                await self.middleware.call('disk.swaps_remove_disks', [p['disk'] for p in mirror['providers']])
+                if mirror_name in existing_swap_devices['mirrors']:
+                    existing_swap_devices['mirrors'].remove(mirror_name)
+            else:
+                if swap_mirror and mirror_name not in existing_swap_devices['mirrors']:
+                    create_swap_devices[mirror_name if IS_LINUX else mirror_name.rsplit('.eli', 1)[0]] = {
```

Comment:
Please group all platform-dependent assignments in one place, it's very hard to understand what's going on now